### PR TITLE
Add Geppetto new automation test framework

### PR DESCRIPTION
**Pull request by mbadrani**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | This is the new test automation framework on the top of dev-tool's chrome protocol.
| Type?         | automation test improvement
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | Follow the steps described on README file.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12069)

### CODE REVIEW

File: tests/Geppetto/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+[![Prestashop](https://i.imgur.com/qDqeQ1E.png)](https://www.prestashop.com)
+
+![Geppetto](./media/logo.png)
+
+# How to Install Geppetto project
+Clone this repo and run them directy with a simple `node` command.
+
+```bash
+git clone https://github.com/mbadrani/geppettov2.git
+cd geppettov2
```
Comment (Quetzacoalt91) [inline tests/Geppetto/README.md]: Doc to be updated with new sub folders

File: tests/Geppetto/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+```
+node test/not_mocha/installPrestashop.js --LANG language --COUNTRY country --DB_SERVER dataBaseServer --DB_USER dataBaseUsername --DB_PASSWD dataBasePassword --RC_TARGET pathOfReleaseTarget --UrlStableVersion urlOfPrestashopRelease --ADMIN_FOLDER_NAME adminFolderName --INSTALL_FOLDER_NAME installFolderName
+```
+
+#### Second example launch PR tests
+
+If you want to run all the prestashop automated Pull Requests tests you can run the campaign **PR**
+
+```
+npm run PR -- --URL='http://127.0.0.1:8081/prestashop' --ADMIN_FOLDER_NAME adminFolderName
```
Comment (Quetzacoalt91) [inline tests/Geppetto/README.md:33]: It seems this script needs to be defined in package.json

```
npm run PR -- --URL='http://ip6-localhost/PrestaShop' --ADMIN_FOLDER_NAME admin-dev
npm ERR! missing script: PR

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/thomas/.npm/_logs/2019-01-08T15_32_00_465Z-debug.log
```

File: tests/Geppetto/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+[![Prestashop](https://i.imgur.com/qDqeQ1E.png)](https://www.prestashop.com)
+
+![Geppetto](./media/logo.png)
+
+# How to Install Geppetto project
+Clone this repo and run them directy with a simple `node` command.
+
+```bash
+git clone https://github.com/mbadrani/geppettov2.git
```
Comment (PierreRambaud) [inline tests/Geppetto/README.md]: hum, the README look outdated :)

File: tests/Geppetto/README.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+[![Prestashop](https://i.imgur.com/qDqeQ1E.png)](https://www.prestashop.com)
```
Comment (PierreRambaud) [inline tests/Geppetto/README.md:1]: Can you rename the directory in lowercase please `gepetto` or better `e2e`.

File: tests/Geppetto/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+```bash
+git clone https://github.com/mbadrani/geppettov2.git
+cd geppettov2
+npm i
+```
+#### First example launch Installation test
+If you want to run the Install test you can run the script **installPrestashop**
+```
+node test/not_mocha/installPrestashop.js --URL UrlOfShop --LANG language --COUNTRY country --DB_SERVER dataBaseServer --DB_USER dataBaseUsername --DB_PASSWD dataBasePassword --ADMIN_FOLDER_NAME adminFolderName --INSTALL_FOLDER_NAME installFolderName
```
Comment (PierreRambaud) [inline tests/Geppetto/README.md]: what is `not_mocha/installPrestashop.js`? Could not find this directory.

File: tests/Geppetto/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Use `TEST_PATH` environment variable to specify which test you want to run.
+
+If you want to run not mocha test only on specific parts (for example PR 8237), you have to run this command:
+
+```
+node test/not_mocha/PR/8237.js --URL 'http://127.0.0.1:8081/prestashop' --ADMIN_FOLDER_NAME adminFolderName
+```
+
+#### The architecture of the project
+![Architecture](./media/tree_archi.png)
```
Comment (PierreRambaud) [inline tests/Geppetto/README.md:37]: The binary is outdated, it should be `campaign`

File: tests/Geppetto/package.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  "version": "1.1.0",
+  "description": "GeppettoV2 is our new version of test Automation Framework for Prestashop",
+  "main": "index.js",
+  "scripts": {
+    "mocha-test-puppeteer": "node_modules/mocha/bin/mocha test/campaigns/clients/take_screenshot.js",
+    "specific-test": "node_modules/mocha/bin/mocha test/campaigns/suite/$TEST_PATH"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/mbadrani/geppettov2.git"
```
Comment (PierreRambaud) [inline tests/Geppetto/package.json:12]: file outdated

File: tests/Geppetto/package.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  "keywords": [
+    "E2E",
+    "test",
+    "automation",
+    "puppetteer"
+  ],
+  "author": "mbadrani",
+  "license": "ISC",
```
Comment (PierreRambaud) [inline tests/Geppetto/package.json:21]: Not sure it is the good license to use. Ping @ttoine

File: tests/Geppetto/package.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "geppettov2",
+  "version": "1.1.0",
+  "description": "GeppettoV2 is our new version of test Automation Framework for Prestashop",
+  "main": "index.js",
+  "scripts": {
+    "mocha-test-puppeteer": "node_modules/mocha/bin/mocha test/campaigns/clients/take_screenshot.js",
```
Comment (PierreRambaud) [inline tests/Geppetto/package.json:7]: rename to `test`

File: tests/Geppetto/package.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  "author": "mbadrani",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/mbadrani/geppettov2/issues"
+  },
+  "homepage": "https://github.com/mbadrani/geppettov2#readme",
+  "dependencies": {
+    "faker": "^4.1.0",
+    "minimist": "^1.2.0",
+    "puppeteer": "^1.10.0"
```
Comment (PierreRambaud) [inline tests/Geppetto/package.json:29]: mocha must be in dependencies and not in dev dependencies. mochawesome too. try to execute the npm command to be sure every dependencies are in the right section :)

File: tests/Geppetto/package.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    "chai": "^4.2.0",
+    "chai-string": "^1.5.0",
+    "eslint-plugin-mocha": "^5.2.0",
+    "file-system": "^2.2.2",
+    "lodash": "^4.17.11",
+    "mocha": "^5.2.0",
+    "mochawesome": "^3.1.1",
+    "should": "^13.2.3"
+  }
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/package.json:41]: `to_populate.js` files must be deleted, they are not needed

File: tests/Geppetto/test/campaigns/clients/common_client.js [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+    await inputFile.uploadFile(fileFolder + fileName);
+  }
+
+  async checkTextValue(selector, textToCheckWith, parameter = 'equal', wait = 0) {
+    switch (parameter) {
+      case "equal":
+        await this.waitFor(wait);
+        await this.waitFor(selector);
+        await page.$eval(selector, el => el.innerText).then((text) => {
+          if (text.indexOf('\t') != -1) {
```
Comment (PierreRambaud) [inline tests/Geppetto/test/campaigns/clients/common_client.js:166]: `'\t'` is different than `"\t"` the test is maybe incorrect.

File: tests/Geppetto/test/campaigns/clients/common_client.js [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+  }
+
+  async switchShopLanguageInFo(language = 'fr') {
+    await this.waitForAndClick(HomePage.language_selector);
+    await this.waitFor(1000);
+    if (language === 'en') {
+      await this.waitForAndClick(HomePage.language_EN);
+    } else {
+      await this.waitForAndClick(HomePage.language_FR);
+    }
```
Comment (PierreRambaud) [inline tests/Geppetto/test/campaigns/clients/common_client.js:203]:
```js
await this.waitForAndClick(language === 'en' ? HomePage.language_EN : HomePage.language_FR);
```

File: tests/Geppetto/test/campaigns/clients/common_client.js [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+    await mouse.move(second_bounding_box.x + (second_bounding_box.width / 2), second_bounding_box.y + (second_bounding_box.height / 2));
+    await mouse.up();
+  }
+
+  async getTextInVar(selector, globalVar, wait = 0, split = false) {
+    if (split) {
+      await this.waitFor(wait);
+      await this.waitFor(selector);
+      await page.$eval(selector, el => el.innerText).then((text) => {
+        global.tab[globalVar] = text.split(': ')[1];
```
Comment (PierreRambaud) [inline tests/Geppetto/test/campaigns/clients/common_client.js:347]: Ony this line change.
You can refactor the entire function by removing duplicate and adding `split ? text.split(': ')[1] : text;`

File: tests/Geppetto/test/campaigns/files/generated_files/download/.gitignore [hunk 15]
```diff
@@ -0,0 +1,3 @@
+*
```
Comment (PierreRambaud) [inline tests/Geppetto/test/campaigns/files/generated_files/download/.gitignore:1]: add an empty `.gitkeep` file instead and configure the .gitignore file in `tests/gepetto`

File: tests/Geppetto/test/campaigns/suite/functional/installationAndUpgrade/01_install.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    scenario('Sign in the "Front Office"', client => {
+      test('should sign in the Front Office', async () => {
+        await client.openShopURL();
+        await client.signInFO();
+      });
+    }, 'common_client');
+  }, 'common_client');
+
+}, 'common_client', true);
+
```
Comment (PierreRambaud) [inline tests/Geppetto/test/campaigns/suite/functional/installationAndUpgrade/01_install.js:24]: too much empty lines

File: tests/Geppetto/test/campaigns/suite/functional/regular/01_install.js [hunk 17]
```diff
@@ -0,0 +1,9 @@
+require('../../functional/installationAndUpgrade/01_install');
+
```
Comment (PierreRambaud) [inline tests/Geppetto/test/campaigns/suite/functional/regular/01_install.js:2]: same here

File: tests/Geppetto/test/globals.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+global.email = argv.LOGIN || 'demo@prestashop.com';
+global.password = argv.PASSWD || 'prestashop_demo';
+
+global.rcTarget = argv.RC_TARGET || '';
+global.rcLink = argv.UrlStableVersion || 'https://download.prestashop.com/download/releases/prestashop_1.7.4.2.zip';
+global.prestashopFolderName = 'prestashop';
+global.URL = argv.URL || 'http://127.0.0.1:8081/prestashop';
+global.language = argv.LANG || 'en';
+global.country = argv.COUNTRY || 'france';
+global.dbServer = argv.DB_SERVER || '10.1.37.2';
```
Comment (PierreRambaud) [inline tests/Geppetto/test/globals.js:16]: set localhost instead

File: tests/Geppetto/test/globals.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+global.dbUser = argv.DB_USER || 'root';
+global.dbPassword = argv.DB_PASSWD || 'sifast2016';
```
Comment (PierreRambaud) [inline tests/Geppetto/test/globals.js:18]: set root instead as default

File: tests/Geppetto/test/helpers.js [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  throw err;
+});
+
+global.test = (name, instructions) => it(name, () => {
+  return instructions().catch(takeScreenshot);
+});
+
+global.scenario = (name, tests, clientName, close = false) =>
+  describe(name, () => {
+    const PrestClient = require("./campaigns/clients/" + clientName);
```
Comment (PierreRambaud) [inline tests/Geppetto/test/helpers.js:16]: use single quote and if possible es6 concatenation instead

File: tests/Geppetto/test/mocha.opts [hunk 21]
```diff
@@ -0,0 +1,4 @@
+--require ./test/helpers.js
+--timeout 400000
+--slow 45000
+--reporter mochawesome --reporter-options reportDir=/projet/presta_test/geppettov2/test/campaigns/files/generated_files/mochawesome-report/
```
Comment (PierreRambaud) [inline tests/Geppetto/test/mocha.opts:4]: wrong reportDir

File: tests/Geppetto/test/uimap/BO/authentication/authentication.js [hunk 22]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  AuthenticationPage: {
+    authentification_page_content_body: 'body', //@TODO
+    authentification_email_input_field: '#email',
+    authentification_password_input_field: '#passwd',
+    authentification_login_button: '#submit_login'
+  }
+};
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/authentication/authentication.js:8]: no newline at the end

File: tests/Geppetto/test/uimap/BO/catalog/products.js [hunk 23]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  ProductPage: {}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/catalog/products.js:2]: should be ProductsPage like others (in plurial)

File: tests/Geppetto/test/uimap/BO/catalog/products.js [hunk 24]
```diff
@@ -0,0 +1,3 @@
...
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/catalog/products.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/customers/customers.js [hunk 25]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  CustomersPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/customers/customers.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/customers/addresses.js [hunk 26]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  AddressesPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/customers/addresses.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/catalog/monitoring.js [hunk 27]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  MonitoringPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/catalog/monitoring.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/catalog/categories.js [hunk 28]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  CategoriesPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/catalog/categories.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/orders/creditSlips.js [hunk 29]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  CreditSlipsPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/orders/creditSlips.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/orders/invoices.js [hunk 30]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  InvoicesPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/orders/invoices.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/BO/orders/orders.js [hunk 31]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  OrdersPage: {}
+}
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/BO/orders/orders.js:3]: no newline at the end

File: tests/Geppetto/test/uimap/installation/install.js [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+      installation_success_populate_database_tables_title: '#process_step_populateDatabase[class="process_step success"]',
+      installation_success_configure_shop_information_title: '#process_step_configureShop[class="process_step success"]',
+      installation_success_install_demonstration_data_title: '#process_step_installFixtures[class="process_step success"]',
+      installation_success_install_modules_title: '#process_step_installModules[class="process_step success"]',
+      installation_success_install_addons_modules_title: '#process_step_installModulesAddons[class="process_step success"]',
+      installation_success_install_theme_title: '#process_step_installTheme[class="process_step success"]',
+      installation_installation_finished_title: '#install_process_success > div:nth-of-type(1) > h2' //@TODO
+    }
+  }
+};
```
Comment (PierreRambaud) [inline tests/Geppetto/test/uimap/installation/install.js:47]: no newline at the end

File: tests/Geppetto/README.md (hunk 2 above)
Comment (mbadrani) [reply tests/Geppetto/README.md:33]: My bad this is due to bad copy paste from my project origin, thanks I will change this

File: tests/Geppetto/README.md (hunk 1 above)
Comment (mbadrani) [reply tests/Geppetto/README.md]: done in my next PR, thanks

File: tests/Geppetto/README.md (hunk 3 above)
Comment (mbadrani) [reply tests/Geppetto/README.md]: fixed in my next PR thanks

File: tests/Geppetto/README.md (hunk 4 above)
Comment (mbadrani) [reply tests/Geppetto/README.md:1]: In fact I've just respected the previous convention, E2E directory has already named with capital letters any suggestion @eternoendless ?

File: tests/Geppetto/README.md (hunk 5 above)
Comment (mbadrani) [reply tests/Geppetto/README.md]: fixed in my next PR thanks

File: tests/Geppetto/package.json (hunk 7 above)
Comment (mbadrani) [reply tests/Geppetto/package.json:12]: fixed in my next PR thanks

File: tests/Geppetto/package.json (hunk 8 above)
Comment (mbadrani) [reply tests/Geppetto/package.json:21]: fixed in my next PR thanks

File: tests/Geppetto/package.json (hunk 9 above)
Comment (mbadrani) [reply tests/Geppetto/package.json:7]: fixed in my next PR thanks

File: tests/Geppetto/package.json (hunk 10 above)
Comment (mbadrani) [reply tests/Geppetto/package.json:29]: fixed in my next PR thanks

File: tests/Geppetto/README.md (hunk 1 above)
Comment (PierreRambaud) [reply tests/Geppetto/README.md]: You don't need to create a new pull request, you can commit directly in the branch.

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
See comments

**Comment by PierreRambaud** [comment]
No activity since one month, a new pull request will be create.

**Pull request closed without merging**